### ISSUE TITLE
[crypto_used_network] use secure protocols for all network communications by default

### ISSUE THREAD

**Issue by cmgrote** [odpi/egeria#1403]
(This is required for **gold** status under CII.)

The software produced by the project MUST support secure protocols for all of its network communications, such as SSHv2 or later, TLS1.2 or later (HTTPS), IPsec, SFTP, and SNMPv3. Insecure protocols such as FTP, HTTP, telnet, SSLv3 or earlier, and SSHv1 MUST be disabled by default, and only enabled if the user specifically configures it.

(see: https://github.com/coreinfrastructure/best-practices-badge/blob/master/doc/other.md#crypto_used_network)

**Comment by bogdan-sava** [comment]
I will enable HTTPS (and disable HTTP) for server-chassis also.
The default used certificate will be by default self signed and included in the source code.

It might affect notebooks and docker images that are using 8080 port and HTTP protocol.

As the UI is using by default 8443 port for HTTPS, in order to be able to run on the same machine my proposal is server-chassis to use by default 9443 port.

**Comment by mandy-chessell** [comment]
Please make sur you include the updates to the tutorials and labs in your PR so there is no disruption to those consuming Egeria.

**Comment by bogdan-sava** [comment]
Of course. Not sure I know all the places where changes are needed, but I'll add you, @planetf1 and @cmgrote as reviewers.

**Comment by planetf1** [comment]
Thanks -- Can I suggest leaving the PR as 'draft' so we don't accidentally merge - as the notebooks etc will definately break without being updated accordingly

**Comment by bogdan-sava** [comment]
Not sure how important is the default self-signed certificate, but I generated default self signed cert that will stay in the resources inside the jar like:

```
 keytool -genkey -alias tomcat -storetype PKCS12 -keyalg RSA -keysize 2048 -keystore keystore.p12 -validity 3650

Enter keystore password:  
Re-enter new password: 
What is your first and last name?
  [Unknown]:  Egeria ODPi
What is the name of your organizational unit?
  [Unknown]:  Egeria
What is the name of your organization?
  [Unknown]:  ODPi
What is the name of your City or Locality?
  [Unknown]:  San Francisco               
What is the name of your State or Province?
  [Unknown]:  CA
What is the two-letter country code for this unit?
  [Unknown]:  US
Is CN=Egeria ODPi, OU=Egeria, O=ODPi, L=San Francisco, ST=CA, C=US correct?
```

**Comment by planetf1** [comment]
A very good question.. I'd be inclined to go with your suggestion.

**Comment by bogdan-sava** [comment]
Changed to CN=*, OU=Egeria, O=ODPi, L=San Francisco, ST=CA, C=US
Java checks the CN match to hostname but allows wildcards

**Comment by bogdan-sava** [comment]
Truststore and cert will be also included.
Unfortunately truststore it is alway interpreted as a filesystem path which makes distribution inside jar not an option and makes distribution difficult.

So making different platforms and UI able to communicate there are following option:

1.  start them with --strict.ssl=false to disable certificate check
2. use the provided truststore with VM options or spring properties like:

`java -Djavax.net.ssl.trustStore=[path/to/truststorer] -Djavax.net.ssl.trustStorePassword=[truststore_pass] -jar server-chassis-[version].jar`

`java -jar server-chassis-[version].jar --server.ssl.trust-store=[path/totruststore] --server.ssl.trust-store-password=[truststore_pass] `

3. override certificates keystore and truststore with prefered ones

4. disable SSL is also possible by using --server.ssl.enable=false

**Comment by planetf1** [comment]
The above focuses on the inbound user requests to server chassis 

We could stop at this point - using http only within the egeria environment (restricting access at the network level) & only have https for consumers -- but this seems very inadequate.  We also have intensive interaction using https between platforms. I wonder if we may need a self-signed CA, and then generate multiple certificates for each server actor? (as well as documenting how to configure them for a production environment)

Each time egeria makes a connection we need to ensure that supports/is configurable to use SSL with appropriate certs

* We've already explored SSL for kafka (configuration, supported in vdc chart)
* graph db / janus / cassandra
* ldap, ranger ....

In terms of the current container setup, ultimately it should support security. STRICT_SSL works well for disabling cert checking   - though a further change would be the expected change/addition of port used by server chassis from 8080 to perhaps 8443 - this will mean all the port numbers in our examples, in the Dockerfiles, docker-compose definitions & k8s scripts. Not technically difficult though. Allowing both ports in the medium term makes sense.

We also

**Comment by bogdan-sava** [comment]
By default, all the platforms are using the same self signed certificate, included in jar.

For production env, is a common procedure well known by dev-ops. Every platform needs a keystore and a truststore.

**Comment by planetf1** [comment]
PR #3056 has extensive updates on the current state of testing & what is working so far.

Changes made:
[ ] Egeria server-chassis is now listening on port 9443 
[ ] Self signed cert and truststore added
[ ] Docs updated to new port numbers/https
[ ] postman info added on skipping ssl cert check
[ ] curl/http docs reference how to skip cert check
[ ] tomcat logging added to report listening port on startup
[ ] SSL check bypass added to spring rest connector (non-configurable)
[ ] SSL check bypass added to notebooks & container environment

This first phase is for interoparability - ie ensuring egeria can be used, even if certificate checking is enabled.

[ ] server chassis - REST calls
[ ] swagger UI
[ ] Notebooks
 - [ ] locally
 - [ ] docker-compose
 - [ ] k8s
 - [ ] cts
[ ] VDC chart in k8s
[ ] docker container standalone
[ ] Egeria UI
- [ ] Asset search
- [ ] Lineage UI
- [ ] Subject Area UI
- [ ] Repository Explorer
 -[ ] Type Explorer
[ ] open metadata samples
 - [ ] admin server
 - [ ] discovery service
 - [ ] security 
 - [ ] asset management
 - [ ] governance program
 - [ ] subject area
[ ] Addtional usage within IntelliJ
 - [ ] launching chassis
 - [ ] running a client
[ ] FVT tests
 - [ ] asset consumer
 - [ ] subject area
[ ] Postman scripts
[ ] release notes
[ ] docs


Known Issues
 - Egeria UI self signed certificate not accepted by current version of chrome, safari - works in firefox. This also affects swagger UI
 - Type Explorer fails to connect
 - Subject area OMAS may not be using the standard spring rest connector
 - CTS test fails when doing the 'origin' test as it does not use the spring rest connector
 - All of the other samples except admin currently failing



Once this PR is complete we will continue with proper use/checks of SSL certs IN A NEW PR - which will include removing, or allowing configuration of some bypasses added above.

**Comment by bogdan-sava** [comment]
@planetf1 thanks for the summary, but I don't understand all the Knowns Issues.
What versions of chrome and safari are you referring to, because in my tests all are working (and tested multiple versions of chrome and firefox and one safari). All works fine after accept the certificate. 
Are you sure you are not using a proxy pack from the company in this 2 browsers that blocks the certificate?

My Chrome is:
Google Chrome is up to date
Version 83.0.4103.61 (Official Build) (64-bit)

**Comment by bogdan-sava** [comment]
Certificate validation is not Spring related but JVM related, so I do not understand the statements regarding spring rest connector.
The validation is made by JVM against default truststore of JVM and fails being self-signed. That's why a truststore  was added. 

Default running of the jar
`java -jar egeria-ui-or-chassis .jar`
makes JVM to look into truststore.p12 located into jvm working folder instead of the default jvm cacerts.

Of corse this can be override by using JVM ooptions like:

`java -Djavax.net.ssl.trustStore=/the/path/to/truststore.p12  -Djavax.net.ssl.trustStorePassword=egeria -jar egeria-ui-or-chassis .jar `

**Comment by planetf1** [comment]
On browsers
 - Chrome Version 83.0.4103.61 (Official Build) beta (64-bit) fails 
 - Safari Version 13.1.1 (15609.2.9.1.2) works (with a warning/update to certificates prompted)
 - Safari Release 106 (Safari 13.2, WebKit 15610.1.12.2) fails

Need to look at whether either of the newer browsers are showing bugs (as beta) are are just failing as they are a few weeks ahead of stable -- if so we have a pending issue.

Chrome in particular is very clear:
<img width="742" alt="Screenshot 2020-05-26 at 12 57 11" src="https://user-images.githubusercontent.com/7292002/82898044-82ddde80-9f50-11ea-9084-9719646ba418.png">

**Comment by planetf1** [comment]
Notice that there is no path given for the user to bypass - at least not without manual certificate import (perhaps). This is the primary concern.

Some products are now generating keys dynamically on install to avoid this issue - but that's a whole lot more infrastructure.

**Comment by planetf1** [comment]
Ref certificate validation - in the clients this is related to the behaviour of restTemplate in particular. I am working with each of our code samples to ensure they work, but need to do more debugging

Limited resource this week due to our Egeria dojo sessions

**Comment by planetf1** [comment]
Type explorer working (https://localhost:9444 / cocoMDS1) + glossary authoring UI

**Comment by planetf1** [comment]
Discovery service working ok (from notebook sample)

**Comment by planetf1** [comment]
asset consumer fvt is an empty module... so no tests. Will leave in-situ as we hope to expand the variety of tests

**Comment by planetf1** [comment]
Fixed subject area FVT to work in std coco config & tested with SSL

**Comment by planetf1** [comment]
Work for the next PR
- [ ] configurations of ssl certs for clients (docs, code)
- [ ] update notebook python code to fully support ssl certs
- [ ] remove SSL workarounds (cts, sa omas, spring rest connector)
- [ ] Release notes

**Comment by planetf1** [comment]
Having checked the security officer omas
 - there's no current tests/demo for this omas
 - this area needs some refactoring (ranger)
 - the correct change is to replace with spring rest connector - and should be done in the next prior before we 'complete' this work - so we're only concerned on the current master 
 - The change already made in the above should fix it for all subsequent useage of restTemplate (as with cts) once the platform is running.
Therefore will deem this change complete (good enough)

**Comment by planetf1** [comment]
Not updating screenshots for this intermediate update as we may need to do again once we have security certs properly checked. Added to todo list for evaluating next set of changes

**Comment by planetf1** [comment]
Autoclosed in error.

The first part is now merged to master. Continuing with part 2 - will move todo list below:

**Comment by planetf1** [comment]
Work for the next PR - or new issues. Keeping as a todo list.

I will likely make unique issues out of some of these as easier to track across multiple owners

- [ ] Investigate if certificates could be per-virtual server or only by platform - ie agree granularity of certificates used
- [ ] Support configuration of certificates & keystore used (clients & server)
- [ ] Update cts, subject area omas (#3112), security officer omas to use spring rest connector
- [ ] Consider #1087 (moving away from restTemplate)
- [ ] replace restTemplate SSL workarounds with configurable support for skipping check / allowing check of certs (consider spring profile)
- [ ] update notebook python code to fully support ssl certs - configuration & skipping CA check
- [ ] Consolidate multiple copies of certificate (build process ?)
- [ ] Update certificate, checkin & Add info on certificate generation into build
- [ ] Access notebook via ssl
- [ ] LDAP lookup from UI via ssl - verify
- [ ] Ranger plugin, ranger, gaian, atlas via ssl
- [ ] cassandra connector/ssl
- [ ] Janus server/ssl
- [ ] presentation server configuration (#3098)
- [ ] Documentation (configuration, defaults, troubleshooting, ports, client, server, notebooks, example using letsencrypt..., curl, httpie, postman)
- [ ] review security scans
- [ ] update screenshots
- [ ] Release notes
- [ ] add note to dockerhub to document the change
- [ ] test in cloud with real public certificate
- [ ] retest all items (as above)

**Comment by planetf1** [comment]
On #1087 there appears no compelling reason to make this change purely to make the security changes required here easier. For both WebClient & restTemplate in order to modify truststore or change validation we need to define an SSL context, which is then used by a specified httpclient, and this is then passed to the builder for either webClient or restTemplate, thus they appear to operate similarly.

Whilst such a change is more futureproof, it is also another variable and in that regards is probably best done as an independent change. The ssl/http code will be pretty much the same.

Thus we will not make that change here.

**Comment by planetf1** [comment]
**Making socket connections (client role)**

Any client code needs to be able to validate the certificate returned from a server (egeria platform)

Standard java approach for this is to add 
```
-Djavax.net.ssl.trustStore=/path/to/truststore.p12
-Djavax.net.ssl.trustStorePassword=egeria
-Djavax.net/ssl.trustStoreType=p12
```

This should work, but has to be specified on launch.

Note that if we do not configure with the spring mechanism, then this approach should still work

The Standard spring mechanism (after all - it provides restTeamplte) for this might LOOK to be:
```
server.ssl.trust-store=truststore.p12 
server.ssl.trust-store-password=egeria
```

However these are actually only used by a server in validating the client certificates for mutual SSL - they are passed to embedded tomcat. They ARE NOT used for the client checking the authenticity of the server certificates. They are server functionality.  

To specify the the client checks we need to create a custom SSL context & manage the certificates there. 

A variety of approaches seem viable for this including
 * Similar to the two workarounds in place 
 * Implementing RestTemplateCustomizer

As with our workaround - where we implement an all-trusting trust manager, we also here will need to load in the right truststore (from a client perspective) in addition to allowing skipping of ssl checks

The values to be used could likely still be picked up as spring properties - so we might use 'http.client.truststore' or similar to distinguish them from the *server* configuration. Spring has no built in support for this.

(note: Tip for debugging: -Djavax.net.debug=ssl )

At this point we have 1 way SSL. The server is presenting a certificate, which the client is validating (or not) but the client is not presenting a cert to the server. 

Adding this moves us to Mutual (2 way) TLS. EACH CLIENT would have it's own certificate, which means plenty to manage…

This is where the spring values above become relevant ie `server.ssl.trust-store` . Whilst there are some reports of this not working, it's possible there was confusion with it's use in restTemplate. By setting the truststore here we are configuring how the server (tomcat) validates client certificates - the opposite way around to the first problem.

So to summarize the planned changes
 - Create a self signed CA
 - Generate a server certificate from the CA
 - Add additional config values as above
 - update SSL context for the client to use these for specifying truststore, changing prior workaround code to use the 'strict.ssl' already in place (may rename) in earlier code [both do the same]
  
Next
 - extend SSL server configuration as needed
 - Generate additional client certificates
 - check mutual TLS

Then
 - Consider if other configuration options beyond spring properties are needed - for example new constructors for spring rest template where the relevant config is passed
 - docker/container/k8s considerations

**Comment by davidradl** [comment]
@planetf1 Some questions so I am on the same page as you on your summary: 
- you say "Create a self signed CA". Where does this live / run ? Is there an option to use an existing CA? Is it one CA per tenant? 
- you say "Generate a server certificate from the CA" . Is this a per server certificate and/or per platform? When does this occur in the lifecycle - I assume this is done after the assembly is deployed, once per server / platform? 
- you say  "update SSL context for the client to use these for specifying truststore,". Does this mean that for 1 way the client needs to be updated with all the server certificates or can it use a CA root certificate?

**Comment by planetf1** [comment]
The creates are one-off (for now), not created dynamically. Just a development activity, only the end result will be added into a deployment -- but we really do need to emphasize this is ONLY to allow a dev to get up and running. They really really really should create a cert/CA at the first opportunity. I will provide sample instructions on doing this, but using normal java/OS tools - nothing egeria specific.

The self signed CA is only to provide something we ship with.  In any real deployment the 'user' would be expected to create any relevant certificates (maybe CAs) and refer to them as we document (tbd). 

Whether they use one CA or multiple is entirely up to them. We will have one server certificate per PLATFORM. we cannot do it at the egeria server level. How that relates to real 'tenants' (as in an org) is deployment specific

The 'update ssl context' is relating to modifying the current code we have in spring rest template. Currently we do fudge the global ssl context, but it's better to do it with a new instance of the SSL context and effectively use that when we construct a restTemplate. The code also needs extending to use the correct truststore.  The client can just use the CA - it doesn't need each certificate. HOWEVER it must also be possible to have a list of certificates in the truststore instead as it will be deployment specific

**Comment by davidradl** [comment]
Hi Nigel,
Ok - I am not too familiar with this - but is sounds reasonable. At the high level what do you think needs to change for the presentation server? I assume I need to do something equivalent to the sslContext change you are making. I am not sure how many scenarios I need to deal with. My guess is I need to deal with the following scenarios: 
* client uses the self signed CA - I know how to do this - if you tell me where it is.
* client uses the truststore? Where will the truststore be, I do not think the presentation server should be accessing content in the ui chassis. I am not sure what this means to the presentation server code.   
* Using an orgs existing CA? Maybe an environment variable specifying the non-standard CA 
* Any other cases?

**Comment by planetf1** [comment]
I’ll work through the base code first part of next week. I suggest we return to this then. I need to limit moving parts in the next step

**Comment by planetf1** [comment]
To elaborate a little more

We have the client and server role - in terms of socket communications - happening in multiple places. For example between an egeria OMAS client and the server chassis, or between a view service and the OMAS client, maybe more levels, then (sorry am not up in full on the UI arch) finally between the user's browser and the web server.

So each of these involves client/server communication, and in each case there is potentially 
 server - side -- a certificate to present 
 client - side - known certs or a CA to check against
AND if using mutual TLS
 server - side - known certs or CA to check the clients against.

So each component, say the view service could be playing both roles in this exchange. Indeed the server chassis itself does, since it also communicates with other metadata repositories.

Generally when we are storing a certificate to send, we use what we call a keystore, whilst when we are doing verification we call it a truststore - but typically it's using the same technology

In my view from reading the spring etc docs & posts there's quite a bit of confusion in this space which I've tried to dig into and clarify. YET at the same time it's new-ish in detail to me, so I may be wrong too.

Therefore I'm planning next to take what I've understood and apply it in the more central levels of the stack - server chassis, omas's and some client samples to prove it.

The UI - both traditional and node-based adds another few layers that won't help the verification, but should be able to follow the same pattern, albeit with different implementatinos due to language.

Hope that explains.

I needed to spend a little time sorting out the release. That has now gone so I expect to progress with some implementation activities at the beginning of next week.

**Comment by davidradl** [comment]
@planetf1 makes sense.

**Comment by planetf1** [comment]
As observed in #3543 we will need to rework the UI accordingly.

In particular a chrome-valid certificate is still needed which should be consolidated with our other certs/docs

**Comment by planetf1** [comment]
I have opened additional issues above to track specific changes we now need moving forward.

The additional configuration added in #3238 may be revisited at a future point, but for now the current behaviour will be documented, including how to use standard java/spring configuration. The main client configuration we do really need is the ability to bypass SSL check, and so there is issue #3869 to do this.

Any other deficiencies/bugs/enhancements should be raised as issues
